src/types/router.rs

```diff
@@ -0,0 +1,630 @@
+        /// The only difference might be how the heuristic function is
+        /// implemented.
+        ///
+        /// # Arguments
+        /// * `from` - The node to start from.
+        /// * `to` - The node to end at.
+        /// * `algorithm` - The algorithm to use.
+        /// * `heuristic` - The heuristic function to use.
+        ///
+        /// # Returns
+        /// A tuple of the total cost and the path consisting of node
+        /// indices.
+        ///
+        /// An empty path with a total cost of 0.0 returned if no path
+        /// is found.
+        ///
+        /// An empty path with a total cost of -1.0 is returned if
+        /// either the `from` or `to` node is not found.
+        pub fn find_shortest_path(
+            &self,
+            from: &Node,
+            to: &Node,
+            algorithm: Algorithm,
+            heuristic_function: Option<fn(NodeIndex) -> f32>,
+        ) -> (f32, Vec<NodeIndex>) {
+            info!(
+                "Finding shortest path from {:?} to {:?} using algorithm {:?}",
+                from.location, to.location, algorithm
+            );
+            if self.get_node_index(from).is_some() && self.get_node_index(to).is_some() {
+                let from_index = self.get_node_index(from).unwrap();
+                let to_index = self.get_node_index(to).unwrap();
+                match algorithm {
+                    Algorithm::Dijkstra => astar(
+                        &self.graph,
+                        from_index,
+                        |finish| finish == to_index,
+                        |e| (*e.weight()).into_inner(),
+                        heuristic_function.unwrap_or(|_| 0.0),
+                    )
+                    .unwrap_or((0.0, Vec::new())),
+
+                    Algorithm::AStar => astar(
+                        &self.graph,
+                        from_index,
+                        |finish| finish == to_index,
+                        |e| (*e.weight()).into_inner(),
+                        heuristic_function.unwrap_or(|_| 0.0),
+                    )
+                    .unwrap_or((0.0, Vec::new())),
+                }
+            } else {
+                warn!("Either the from or to node is not found.");
+                (-1.0, Vec::new())
+            }
+        }
+
+        /// Compute the total Haversine distance of a path.
+        ///
+        /// # Arguments
+        /// * `path` - The path to compute the distance of. The path is
+        ///   given as a vector of [`NodeIndex`] structs.
+        ///
+        /// # Returns
+        /// The total distance of the path.
+        ///
+        /// If the path is empty, 0.0 is returned.
+        ///
+        /// If the path is invalid, -1.0 is returned.
+        pub fn get_total_distance(&self, path: &Vec<NodeIndex>) -> StdResult<f32, RouterError> {
+            info!("Computing total distance of path");
+            let mut total_distance = 0.0;
+            for i in 0..path.len() - 1 {
```

Comment:
Do you need the  "- 1"?

0..path.len() equates to 0..9 if path.len() is 10

0..=path.len() equates to 0..10

Reply:
ah nevermind, I see you do path[i + 1] in the body.